File: dummy-apps/ruby-rails-react/backend/app/models/tax.rb
class Tax < ApplicationRecord
  STATUSES = %w[draft filed paid overdue].freeze
  FILING_STATUSES = %w[single married_jointly married_separately head_of_household].freeze

  validates :tax_year, presence: true, numericality: { only_integer: true, greater_than_or_equal_to: 1900 }
  validates :jurisdiction, presence: true
  validates :filing_status, presence: true, inclusion: { in: FILING_STATUSES }
  validates :currency, presence: true, length: { is: 3 }
  validates :status, presence: true, inclusion: { in: STATUSES }

  validates :income, :deductions, :taxable_income, :tax_amount,
            numericality: { greater_than_or_equal_to: 0 }
end

File: dummy-apps/ruby-rails-react/frontend/src/types/tax.ts
export type Tax = {
  id: number;
  tax_year: number;
  jurisdiction: string;
  filing_status: string;
  income: number;
  deductions: number;
  taxable_income: number;
  tax_amount: number;
  currency: string;
  status: string;
  due_date?: string;
  paid_at?: string | null;
  notes?: string | null;
  metadata: Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
};

export type TaxListResponse = {
  data: Tax[];
  meta?: {
    total?: number;
    limit?: number;
    offset?: number;
  };
};
